src/NexusMods.Library/AddLibraryFileJobWorker.cs

```diff
@@ -0,0 +1,218 @@
+using JetBrains.Annotations;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using NexusMods.Abstractions.FileExtractor;
+using NexusMods.Abstractions.IO.StreamFactories;
+using NexusMods.Abstractions.Jobs;
+using NexusMods.Abstractions.Library;
+using NexusMods.Hashing.xxHash64;
+using NexusMods.MnemonicDB.Abstractions;
+using NexusMods.Paths;
+
+namespace NexusMods.Library;
+
+[UsedImplicitly]
+internal class AddLibraryFileJobWorker : AJobWorker<AddLibraryFileJob>
+{
+    private readonly ILogger _logger;
+
+    private readonly IServiceProvider _serviceProvider;
+    private readonly IFileExtractor _fileExtractor;
+    private readonly TemporaryFileManager _temporaryFileManager;
+
+    public AddLibraryFileJobWorker(IServiceProvider serviceProvider)
+    {
+        _logger = serviceProvider.GetRequiredService<ILogger<AddLibraryFileJobWorker>>();
+
+        _serviceProvider = serviceProvider;
+        _fileExtractor = serviceProvider.GetRequiredService<IFileExtractor>();
+        _temporaryFileManager = serviceProvider.GetRequiredService<TemporaryFileManager>();
+    }
+
+    protected override async Task<JobResult> ExecuteAsync(AddLibraryFileJob job, CancellationToken cancellationToken)
+    {
+        var absolutePath = job.FilePath;
+        if (!absolutePath.FileExists)
+        {
+            if (absolutePath.DirectoryExists())
+            {
+                return JobResult.CreateFailed($"File at `{absolutePath}` can't be added to the library because it's a directory");
+            }
+
+            return JobResult.CreateFailed($"File at `{absolutePath}` can't be added to the library because it doesn't exist");
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+        if (!job.IsArchive.HasValue)
+        {
+            job.IsArchive = await CheckIfArchiveAsync(job.FilePath);
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+        if (!job.HashJobResult.HasValue)
+        {
+            job.HashJobResult = await HashAsync(job.FilePath, cancellationToken);
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+        if (!job.EntityId.HasValue)
+        {
+            job.EntityId = job.Transaction.TempId();
+        }
+
```

Comment:
I don't really like this list of ifs to keep track of current resume state.
Have you considered using a switch and an internal status value to keep track of the step we are on?
Similar to async tasks state machines?

Reply:
Can you provide an example of what you might want to see?

Reply:
We are mixing data validation with resume logic, since there is a lot of overlap, but it does mean that attempting to resume the process after pausing would go through every step, requiring the keeping of all intermediate steps data, instead of just the data required for the next step. 

A switch would allow jumping directly to the last completed state.
This is actually made worse due to C# not doing cascading fall through of positive switch cases like other languages do.
```csharp
protected override async Task<JobResult> ExecuteAsync(AddLibraryFileJob job, CancellationToken cancellationToken)
{
    var absolutePath = job.FilePath;

    switch (job.LastCompletedStep)
    {
        case AddLibraryFileJobStep.Start:
            goto CheckIfArchive;
        case AddLibraryFileJobStep.CheckIfArchive:
            goto Hash;
        case AddLibraryFileJobStep.Hash:
            goto CreateLibraryFile;
        case AddLibraryFileJobStep.CreateLibraryFile:
            goto CreateLibraryArchive;
        case AddLibraryFileJobStep.CreateLibraryArchive:
            goto ExtractArchive;
        case AddLibraryFileJobStep.ExtractArchive:
            goto AddExtractedFiles;
        case AddLibraryFileJobStep.AddExtractedFiles:
            goto Commit;
        case AddLibraryFileJobStep.Commit:
        case AddLibraryFileJobStep.Completed:
            return JobResult.CreateCompleted(job.LibraryFile.Value);
    }

CheckIfArchive:
    // Your existing logic for checking if archive
    job.LastCompletedStep = AddLibraryFileJobStep.CheckIfArchive;
    cancellationToken.ThrowIfCancellationRequested();

Hash:
    // Your existing logic for hashing
    job.LastCompletedStep = AddLibraryFileJobStep.Hash;
    cancellationToken.ThrowIfCancellationRequested();

CreateLibraryFile:
    // Your existing logic for creating library file
    job.LastCompletedStep = AddLibraryFileJobStep.CreateLibraryFile;
    cancellationToken.ThrowIfCancellationRequested();

CreateLibraryArchive:
    // Your existing logic for creating library archive if necessary
    job.LastCompletedStep = AddLibraryFileJobStep.CreateLibraryArchive;
    cancellationToken.ThrowIfCancellationRequested();

ExtractArchive:
    // Your existing logic for extracting archive if necessary
    job.LastCompletedStep = AddLibraryFileJobStep.ExtractArchive;
    cancellationToken.ThrowIfCancellationRequested();
```

Reply:
That doesn't really seem easier to use.

Reply:
This is not supposed to be a persistent job right? If the job is paused, it would be lost on app close right?